### PR TITLE
refactor: refactor x/exchange module

### PR DESCRIPTION
**Pull request by kingcre**
## Description

- Use `sdk.Int` for quantities
- Simplify fee calculation and ceiling/truncation for order amounts
- Move `min_order_quantity`, `min_order_quote` from x/amm `Pool` into x/exchange `Market` and also add `max_order_quantity`, `max_order_quote`
- Move fees and quote coin dust into `Market`'s `fee_collector`
- Change `MaxPrice` to 10^24
- Store price as tick indexes

### CODE REVIEW

File: proto/crescent/amm/v1beta1/params.proto [hunk 1]
```diff
@@ -12,11 +12,7 @@ option (gogoproto.goproto_getters_all) = false;
...
       [(gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins", (gogoproto.nullable) = false];
-  uint32 default_tick_spacing       = 2;
-  string default_min_order_quantity = 3
-      [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
-  string default_min_order_quote = 4
-      [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
+  uint32   default_tick_spacing                                       = 2;
   repeated cosmos.base.v1beta1.Coin private_farming_plan_creation_fee = 5
       [(gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins", (gogoproto.nullable) = false];
   uint32                   max_num_private_farming_plans = 6;
```
Comment (blackhatcre) [inline proto/crescent/amm/v1beta1/params.proto]:
```suggestion
  repeated cosmos.base.v1beta1.Coin private_farming_plan_creation_fee = 3
      [(gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins", (gogoproto.nullable) = false];
  uint32                   max_num_private_farming_plans = 4;
```

File: app/upgrades/mainnet/v5/upgrade.go [hunk 2]
```diff
@@ -70,11 +70,11 @@ func UpgradeHandler(
 			sdk.NewDecWithPrec(1, 3), // Maker: 0.1%
 			sdk.NewDecWithPrec(2, 3), // Taker: 0.2%
 			sdk.NewDecWithPrec(8, 1)) // Order source: Taker * 80%
+		exchangeParams.DefaultMinOrderQuantity = sdk.NewInt(10000)
+		exchangeParams.DefaultMinOrderQuote = sdk.NewInt(10000)
```
Comment (blackhatcre) [inline app/upgrades/mainnet/v5/upgrade.go]: No need to add `DefaultMaxOrderQuantity` and `DefaultMinOrderQuote`?

File: app/upgrades/mainnet/v5/upgrade.go [hunk 3]
```diff
@@ -573,6 +575,12 @@ func UpgradeHandler(
 				if ParamChanges[pairId].TakerFeeRate != nil {
 					market.TakerFeeRate = *ParamChanges[pairId].TakerFeeRate
 				}
+				if ParamChanges[pairId].MinOrderQuantity != nil {
+					market.MinOrderQuantity = *ParamChanges[pairId].MinOrderQuantity
+				}
+				if ParamChanges[pairId].MinOrderQuote != nil {
+					market.MinOrderQuote = *ParamChanges[pairId].MinOrderQuote
+				}
```
Comment (blackhatcre) [inline app/upgrades/mainnet/v5/upgrade.go]: ditto

File: proto/crescent/exchange/v1beta1/event.proto [hunk 4]
```diff
@@ -94,26 +94,22 @@ message EventCancelAllOrders {
...
-  string price    = 5 [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
-  string quantity = 6 [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
-  string open_quantity = 7
-      [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
-  string executed_quantity = 8
-      [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
-  cosmos.base.v1beta1.DecCoin paid     = 9 [(gogoproto.nullable) = false];
-  cosmos.base.v1beta1.DecCoin received = 10 [(gogoproto.nullable) = false];
+  uint64 order_id = 1;
+  string quantity = 2 [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int", (gogoproto.nullable) = false];
```
Comment (blackhatcre) [inline proto/crescent/exchange/v1beta1/event.proto:113]: I think the `market_id`, `orderer`, `is_buy`, and `price` are necessary or useful as fields for filled orders event. Why did you remove them?

File: x/amm/keeper/source.go [hunk 5]
```diff
@@ -127,8 +127,8 @@ func (k Keeper) AfterPoolOrdersExecuted(ctx sdk.Context, pool types.Pool, result
 		extraAmt0 = utils.ZeroDec
 		extraAmt1 = utils.ZeroDec
 	}
-	for i, result := range results {
-		orderTick := exchangetypes.TickAtPrice(result.Price())
+	for i, order := range results {
```
Comment (blackhatcre) [inline x/amm/keeper/source.go]: How about replace argument `results` to `orders`?

File: x/exchange/keeper/proposal_handler.go [hunk 6]
```diff
@@ -16,12 +16,20 @@ func HandleMarketParameterChangeProposal(ctx sdk.Context, k Keeper, p *types.Mar
 		market.MakerFeeRate = change.MakerFeeRate
 		market.TakerFeeRate = change.TakerFeeRate
 		market.OrderSourceFeeRatio = change.OrderSourceFeeRatio
+		if change.MinOrderQuantity != nil {
+			market.MinOrderQuantity = *change.MinOrderQuantity
+		}
+		if change.MinOrderQuote != nil {
+			market.MinOrderQuote = *change.MinOrderQuote
+		}
```
Comment (blackhatcre) [inline x/exchange/keeper/proposal_handler.go]: Please add `MaxOrderQuantity` and `MaxOrderQuote`.

File: x/exchange/keeper/proposal_handler.go [hunk 7]
```diff
@@ -16,12 +16,20 @@ func HandleMarketParameterChangeProposal(ctx sdk.Context, k Keeper, p *types.Mar
...
 		k.SetMarket(ctx, market)
 		if err := ctx.EventManager().EmitTypedEvent(&types.EventMarketParameterChanged{
 			MarketId:            change.MarketId,
 			MakerFeeRate:        change.MakerFeeRate,
 			TakerFeeRate:        change.TakerFeeRate,
 			OrderSourceFeeRatio: change.OrderSourceFeeRatio,
+			MinOrderQuantity:    change.MinOrderQuantity,
+			MinOrderQuote:       change.MinOrderQuote,
```
Comment (blackhatcre) [inline x/exchange/keeper/proposal_handler.go]: ditto

File: x/exchange/simulation/proposals.go [hunk 8]
```diff
@@ -35,7 +35,9 @@ func SimulateMarketParameterChangeProposal(k keeper.Keeper) simtypes.ContentSimu
 			if r.Float64() <= 0.5 {
 				fees := GenFees(r)
 				changes = append(changes, types.NewMarketParameterChange(
-					market.Id, fees.DefaultMakerFeeRate, fees.DefaultTakerFeeRate, fees.DefaultOrderSourceFeeRatio))
+					market.Id,
+					fees.DefaultMakerFeeRate, fees.DefaultTakerFeeRate, fees.DefaultOrderSourceFeeRatio,
+					nil, nil, nil, nil)) // XXX
```
Comment (blackhatcre) [inline x/exchange/simulation/proposals.go]: Please remove comment that seems useless. If it is useful, replace it with something else.

File: x/exchange/types/proposal.go [hunk 9]
```diff
@@ -89,5 +99,29 @@ func (change MarketParameterChange) Validate() error {
...
+			return sdkerrors.Wrapf(
+				sdkerrors.ErrInvalidRequest, "max order quantity must not be negative: %s", change.MaxOrderQuantity)
+		}
+	}
+	if change.MaxOrderQuote != nil {
+		if change.MaxOrderQuote.IsNegative() {
+			return sdkerrors.Wrapf(
+				sdkerrors.ErrInvalidRequest, "max order quote must not be negative: %s", change.MaxOrderQuote)
+		}
+	}
```
Comment (blackhatcre) [inline x/exchange/types/proposal.go]: For both `Quantity` and `Quote`, I think it is better to test if `MinOrderXXX` <= `MaxOrderXXX`, if `MinOrderXXX` and `MaxOrderXXX` are not nil.

File: x/exchange/types/market.go [hunk 10]
```diff
@@ -42,17 +53,56 @@ func (market Market) Validate() error {
...
+	}
+	if market.MinOrderQuote.IsNegative() {
+		return fmt.Errorf("min order quote must not be negative: %s", market.MinOrderQuote)
+	}
+	if market.MaxOrderQuantity.IsNegative() {
+		return fmt.Errorf("max order quantity must not be negative: %s", market.MaxOrderQuantity)
+	}
+	if market.MaxOrderQuote.IsNegative() {
+		return fmt.Errorf("max order quote must not be negative: %s", market.MaxOrderQuote)
+	}
```
Comment (blackhatcre) [inline x/exchange/types/market.go]: How about add test if `MinOrderXXX` <= `MaxOrderXXX`?

File: x/exchange/types/keys.go [hunk 11]
```diff
@@ -146,13 +146,31 @@ func ParseNumMMOrdersKey(key []byte) (ordererAddr sdk.AccAddress, marketId uint6
...
+	if !valid {
+		panic("invalid tick price")
+	}
+	var signByte []byte
+	if tick < 0 {
+		signByte = zeroByte
+	} else {
+		signByte = oneByte
+	}
+	return utils.Key(signByte, utils.Uint32ToBigEndian(uint32(tick)))
```
Comment (blackhatcre) [inline x/exchange/types/keys.go:173]: I cannot find a role for `sign byte`, is there a reason to store it?

File: x/exchange/keeper/genesis_test.go [hunk 12]
```diff
@@ -13,8 +13,8 @@ func (s *KeeperTestSuite) TestImportExportGenesis() {
 	s.CreateMarket("ucre", "uusd")
 	ordererAddr1 := s.FundedAccount(1, enoughCoins)
 	ordererAddr2 := s.FundedAccount(1, enoughCoins)
-	s.PlaceLimitOrder(1, ordererAddr1, true, utils.ParseDec("4.9"), sdk.NewDec(10_000000), time.Hour)
-	s.PlaceLimitOrder(1, ordererAddr2, false, utils.ParseDec("5"), sdk.NewDec(20_000000), time.Hour)
+	s.PlaceLimitOrder(1, ordererAddr1, true, utils.ParseDec("4.9"), sdk.NewInt(10_000000), time.Hour)
+	s.PlaceLimitOrder(1, ordererAddr2, false, utils.ParseDec("5"), sdk.NewInt(20_000000), time.Hour)
```
Comment (richard-bachman) [inline x/exchange/keeper/genesis_test.go:18]: Can the test lines for MM orders, market orders, and batch orders be possible?

File: x/exchange/keeper/market.go [hunk 13]
```diff
@@ -28,9 +28,14 @@ func (k Keeper) CreateMarket(
...
+	defaultMaxOrderQty := k.GetDefaultMaxOrderQuantity(ctx)
+	defaultMaxOrderQuote := k.GetDefaultMaxOrderQuote(ctx)
 	market = types.NewMarket(
 		marketId, baseDenom, quoteDenom,
-		fees.DefaultMakerFeeRate, fees.DefaultTakerFeeRate, fees.DefaultOrderSourceFeeRatio)
+		fees.DefaultMakerFeeRate, fees.DefaultTakerFeeRate, fees.DefaultOrderSourceFeeRatio,
+		defaultMinOrderQty, defaultMinOrderQuote, defaultMaxOrderQty, defaultMaxOrderQuote)
 	k.SetMarket(ctx, market)
 	k.SetMarketByDenomsIndex(ctx, market)
 	k.SetMarketState(ctx, market.Id, types.NewMarketState(nil))
```
Comment (richard-bachman) [inline x/exchange/keeper/market.go:38]: In the function of `EscrowCoins` and `ReleaseCoins`, the checking to see if `coins` is all zero might be needed.

File: x/exchange/keeper/order.go [hunk 14]
```diff
@@ -110,7 +109,7 @@ func (k Keeper) PlaceMMBatchLimitOrder(
 
 func (k Keeper) placeLimitOrder(
 	ctx sdk.Context, typ types.OrderType, marketId uint64, ordererAddr sdk.AccAddress,
-	isBuy bool, price, qty sdk.Dec, lifespan time.Duration, isBatch bool) (orderId uint64, order types.Order, res types.ExecuteOrderResult, err error) {
+	isBuy bool, price sdk.Dec, qty sdk.Int, lifespan time.Duration, isBatch bool) (orderId uint64, order types.Order, res types.ExecuteOrderResult, err error) {
 	if !qty.IsPositive() { // sanity check
 		panic("quantity must be positive")
 	}
```
Comment (richard-bachman) [inline x/exchange/keeper/order.go:130]: How about adding one sanity check below?
```
if !price.IsPositive() { // HERE!! sanity check
	panic("price must be positive")
}
```

File: x/exchange/keeper/order.go [hunk 15]
```diff
@@ -120,6 +119,32 @@ func (k Keeper) placeLimitOrder(
...
+			qty, market.MinOrderQuantity)
+		return
+	} else if qty.GT(market.MaxOrderQuantity) {
+		err = sdkerrors.Wrapf(
+			sdkerrors.ErrInvalidRequest,
+			"quantity is greater than the maximum order quantity allowed: %s > %s",
+			qty, market.MaxOrderQuantity)
+		return
+	}
+	if quote := price.MulInt(qty).TruncateInt(); quote.LT(market.MinOrderQuote) {
```
Comment (richard-bachman) [inline x/exchange/keeper/order.go]: What about having `Ceil()` for safety?

File: x/exchange/keeper/proposal_handler_test.go [hunk 16]
```diff
@@ -35,7 +36,8 @@ func (s *KeeperTestSuite) TestPoolParameterChangeProposal() {
 	proposal = types.NewMarketParameterChangeProposal(
 		"Title", "Description", []types.MarketParameterChange{
 			types.NewMarketParameterChange(
-				3, utils.ParseDec("0.001"), utils.ParseDec("0.002"), utils.ParseDec("0.5")),
+				3, utils.ParseDec("0.001"), utils.ParseDec("0.002"), utils.ParseDec("0.5"),
+				nil, nil, nil, nil),
 		})
 	s.Require().NoError(proposal.ValidateBasic())
 	s.Require().EqualError(handler(s.Ctx, proposal), "market 3 not found: not found")
```
Comment (richard-bachman) [inline x/exchange/keeper/proposal_handler_test.go:54]: How about having one more test as below?
```
func (s *KeeperTestSuite) TestPoolParameterChangeProposal_WrongFees() {
 ...
}
```

File: x/exchange/keeper/swap_test.go [hunk 17]
```diff
@@ -9,6 +9,37 @@ import (
 	"github.com/crescent-network/crescent/v5/x/exchange/types"
 )
```
Comment (richard-bachman) [inline x/exchange/keeper/swap_test.go:11]: What about having the following tests?
```
// HERE! Test for the halves fee
func (s *KeeperTestSuite) TestSwapExactAmountIn_HalveFee() {

}

// HERE! Test for FindAllRoutes
func (s *KeeperTestSuite) TestFindAllRoutes() {
	// Add markets: ucre/uusd, ucre/uatom, uusd/uatom, uusd/ustake, uatom/ustake

}

// HERE! Test for FindAllRoutes with no last prices markets
func (s *KeeperTestSuite) TestFindAllRoutes_NoLastPrice() {
	// Add markets: ucre/uusd, ucre/uatom, uusd/uatom, uusd/ustake, uatom/ustake

}

```

File: x/exchange/keeper/swap.go [hunk 18]
```diff
@@ -68,18 +68,17 @@ func (k Keeper) SwapExactAmountIn(
...
 			MarketId:         marketId,
 			ExecutedQuantity: res.ExecutedQuantity,
 			Input:            currentIn,
 			Output:           output,
-			Fee:              res.Fee,
+			// TODO: add Paid field?
+			Fee:              res.FeePaid,
 		})
 		currentIn = output
 	}
```
Comment (richard-bachman) [inline x/exchange/keeper/swap.go:89]: In Line 118, what if the market has no order? Do we need to skip it if the market has no order?

File: x/exchange/types/escrow.go [hunk 19]
```diff
@@ -38,9 +38,9 @@ func (e *Escrow) Unlock(addr sdk.AccAddress, amt ...sdk.DecCoin) {
...
 func (e *Escrow) Pays(addr sdk.AccAddress) sdk.Coins {
 	var pays sdk.Coins
-	for _, decCoin := range e.deltas[addr.String()] {
-		if decCoin.IsNegative() {
-			coin := sdk.NewCoin(decCoin.Denom, decCoin.Amount.Neg().Ceil().TruncateInt())
+	for _, coin := range e.deltas[addr.String()] {
+		if coin.IsNegative() {
+			coin.Amount = coin.Amount.Neg()
 			pays = pays.Add(coin)
 		}
```
Comment (richard-bachman) [inline x/exchange/types/escrow.go]: What about the case of IsPositive()? If it is not allowed, the error handling (or sanity check) is needed.

File: x/exchange/types/escrow.go [hunk 20]
```diff
@@ -49,9 +49,8 @@ func (e *Escrow) Pays(addr sdk.AccAddress) sdk.Coins {
 
 func (e *Escrow) Receives(addr sdk.AccAddress) sdk.Coins {
 	var receives sdk.Coins
-	for _, decCoin := range e.deltas[addr.String()] {
-		if decCoin.IsPositive() {
-			coin, _ := decCoin.TruncateDecimal()
+	for _, coin := range e.deltas[addr.String()] {
+		if coin.IsPositive() {
 			receives = receives.Add(coin)
 		}
```
Comment (richard-bachman) [inline x/exchange/types/escrow.go]: What about the case of IsNegative()? If it is not allowed, the error handling (or sanity check) is needed.

File: x/exchange/types/market.go [hunk 21]
```diff
@@ -42,17 +53,56 @@ func (market Market) Validate() error {
...
 	return nil
```
Comment (richard-bachman) [inline x/exchange/types/market.go:64]: Are the following two checks needed?
```
// HERE! min order quantity <= max order quantity
	if market.MinOrderQuantity.GT(market.MaxOrderQuantity) {
		return fmt.Errorf(
			"min order quantity must not be greater than max order quantity: %s, %s",
			market.MinOrderQuantity, market.MaxOrderQuantity)
	}
	// HERE! min order quote <= max order quote
	if market.MinOrderQuote.GT(market.MaxOrderQuote) {
		return fmt.Errorf(
			"min order quote must not be greater than max order quote: %s, %s",
			market.MinOrderQuote, market.MaxOrderQuote)
	}
```

File: x/exchange/types/matching_test.go [hunk 22]
```diff
@@ -23,21 +22,27 @@ func TestPayReceiveDenoms(t *testing.T) {
 
 func TestFillMemOrderBasic(t *testing.T) {
 	market := types.NewMarket(
-		1, "ucre", "uusd", utils.ParseDec("-0.0015"), utils.ParseDec("0.003"), utils.ParseDec("0.5"))
+		1, "ucre", "uusd",
+		utils.ParseDec("-0.0015"), utils.ParseDec("0.003"), utils.ParseDec("0.5"),
```
Comment (richard-bachman) [inline x/exchange/types/matching_test.go]: Can we have a negative fee rate?

File: x/exchange/types/matching_test.go [hunk 23]
```diff
@@ -23,21 +22,27 @@ func TestPayReceiveDenoms(t *testing.T) {
...
+	ctx.FillOrder(order, sdk.NewInt(5500), utils.ParseDec("1.0013"), false)
+	res := order.Result()
+	utils.AssertEqual(t, sdk.NewInt(5500), res.ExecutedQuantity)
+	utils.AssertEqual(t, sdk.NewInt(5508), res.Paid)
+	utils.AssertEqual(t, sdk.NewInt(5483), res.Received)
+	utils.AssertEqual(t, sdk.NewInt(17), res.FeePaid)
+	utils.AssertEqual(t, sdk.NewInt(0), res.FeeReceived)
+
+	order = newUserMemOrder(2, false, utils.ParseDec("1.0015"), sdk.NewInt(10000), sdk.NewInt(9000))
+	ctx.FillOrder(order, sdk.NewInt(5500), utils.ParseDec("1.0013"), true)
```
Comment (richard-bachman) [inline x/exchange/types/matching_test.go:58]: For sell order, the ordered price should be lower than the filled price.
But it seems that the ordered price (1.0015) is larger than the filled price (1.0013).

File: x/exchange/types/matching.go [hunk 24]
```diff
@@ -29,62 +119,35 @@ func NewMatchingContext(market Market, halveFees bool) *MatchingContext {
...
-		pays = pays.Add(fee)
-	} else if feeRate.IsPositive() {
-		receives, fee = DeductFee(receives, feeRate)
-	} else {
-		fee = utils.ZeroDec
+	return feeRate
+}
+
+func (ctx *MatchingContext) FillOrder(order *MemOrder, qty sdk.Int, price sdk.Dec, isMaker bool) {
+	executableQty := order.ExecutableQuantity()
```
Comment (richard-bachman) [inline x/exchange/types/matching.go:145]: For sanity check, can we have checks to see if order.Price >= price for buy order and order.Price <= price for sell order?

File: x/exchange/types/matching.go [hunk 25]
```diff
@@ -164,76 +227,68 @@ func (ctx *MatchingContext) MatchOrderBookPriceLevels(
...
 		}
 		if quoteLimit != nil {
-			remainingQuote := quoteLimit.Sub(res.ExecutedQuote)
-			if remainingQuote.LT(utils.OneDec) {
-				res.FullyExecuted = true
-				break
-			}
-			qty := remainingQuote.QuoTruncate(level.price)
+			remainingQuote = quoteLimit.ToDec().Sub(totalExecutedQuote)
+			qty := remainingQuote.QuoTruncate(level.Price).TruncateInt()
```
Comment (richard-bachman) [inline x/exchange/types/matching.go:258]: Do we need to consider isBuy for TruncateInt() ceiling or flooring?

File: x/exchange/types/mem_order_test.go [hunk 26]
```diff
@@ -76,28 +76,28 @@ func TestMemOrder_HasPriorityOver(t *testing.T) {
...
+	order2 = newUserMemOrder(2, false, price, sdk.NewInt(100_000000), sdk.NewInt(90_000000))
 	// lower order id has priority
 	require.True(t, order1.HasPriorityOver(order2))
 
-	order1 = newUserMemOrder(1, false, price, sdk.NewDec(100_000000), sdk.NewDec(90_000000))
-	order2 = newOrderSourceMemOrder(false, price, sdk.NewDec(100_000000), source1)
+	order1 = newUserMemOrder(1, false, price, sdk.NewInt(100_000000), sdk.NewInt(90_000000))
+	order2 = newOrderSourceMemOrder(false, price, sdk.NewInt(100_000000), source1)
 	// OrderSourceMemOrder has priority over UserMemOrder
 	require.True(t, order2.HasPriorityOver(order1))
```
Comment (richard-bachman) [inline x/exchange/types/mem_order_test.go:94]: The following can be added.
```
require.False(t, order1.HasPriorityOver(order2))
```

File: x/exchange/types/mem_order.go [hunk 27]
```diff
@@ -29,193 +29,116 @@ func (typ MemOrderType) String() string {
...
-		return sdk.MinDec(executableQty, order.remainingDeposit.QuoTruncate(order.price))
+func (order *MemOrder) ExecutableQuantity() sdk.Int {
+	executableQty := order.OpenQuantity.Sub(order.executedQty)
+	if order.IsBuy {
+		return utils.MinInt(
+			executableQty,
+			order.RemainingDeposit.ToDec().QuoTruncate(order.Price).TruncateInt())
 	}
-	return sdk.MinDec(executableQty, order.remainingDeposit)
+	return utils.MinInt(executableQty, order.RemainingDeposit.ToDec().TruncateInt())
```
Comment (richard-bachman) [inline x/exchange/types/mem_order.go]: Why do we need to truncate after ToDec()? (`RemainingDeposit` is already Int type.)
Can we use the following?
```
return utils.MinInt(executableQty, order.RemainingDeposit)
```

File: x/exchange/types/mem_order.go [hunk 28]
```diff
@@ -29,193 +29,116 @@ func (typ MemOrderType) String() string {
...
-	if !order.price.Equal(other.price) { // sanity check
-		panic(fmt.Sprintf("orders with different price: %s != %s", order.price, other.price))
+	if !order.Price.Equal(other.Price) { // sanity check
+		panic(fmt.Sprintf("orders with different price: %s != %s", order.Price, other.Price))
 	}
-	if !order.qty.Equal(other.qty) {
-		return order.qty.GT(other.qty)
+	if !order.Quantity.Equal(other.Quantity) {
+		return order.Quantity.GT(other.Quantity)
 	}
```
Comment (richard-bachman) [inline x/exchange/types/mem_order.go:105]: Additional condition for sanity check can be possible.
```
if order.IsBuy != other.IsBuy { // sanity check
     panic(fmt.Sprintf("orders with different direction: %t != %t", order.IsBuy, other.IsBuy))
}
```

**Review by richard-bachman — COMMENTED** [review]
Thank you for drafting this PR. 
After reviewing the overall codebase, the followings are the highlights.
- Add test suites and simulations for order source orders and their matching
    - This test for order source orders is very important since this is the part to interaction with other module for order like amm.
- Can we check the matched price and matched amount in `Batch_test`?
    -  Currently, only last price is checked. 
- Much more comments above the functions/variables to explain what they are (even though the explanation is short and brief.
- Please remove all `// TODO: ...` part.
    - Others cannot know if it is not completed parts or to be completed or to be discussed further. 
    - Please make the discussion in order to resolve all `TODO: ` part ASAP.
- What if the max value of life span of limit order?
- Please complete spec.
    - Need to explain in detail about mid-block processing since this is the new concept.
    - `Events`, `README` need to be completed

Other detailed comments are added as inline comments.

File: app/upgrades/mainnet/v5/upgrade.go (hunk 2 above)
Comment (kingcre) [reply app/upgrades/mainnet/v5/upgrade.go]: We can do that, but I think the value we set will be the same as the default value, 10^30.

File: app/upgrades/mainnet/v5/upgrade.go (hunk 3 above)
Comment (kingcre) [reply app/upgrades/mainnet/v5/upgrade.go]: Since we won't change `MaxOrderQuantity` and `MaxOrderQuote` per-pair basis(we'll just use the default value), the code for those is not inlcuded.

File: proto/crescent/amm/v1beta1/params.proto (hunk 1 above)
Comment (kingcre) [reply proto/crescent/amm/v1beta1/params.proto]: Nice, thank you.

File: proto/crescent/exchange/v1beta1/event.proto (hunk 4 above)
Comment (kingcre) [reply proto/crescent/exchange/v1beta1/event.proto:113]: Yes those fields are useful, but those don't change over time so I thought emitting it every time the order is matched is a bit redundant. Someone who need to figure out those fields can always reference the order by the id.

File: x/amm/keeper/source.go (hunk 5 above)
Comment (kingcre) [reply x/amm/keeper/source.go]: It's a good point. I'll rename it.

File: x/exchange/keeper/proposal_handler.go (hunk 6 above)
Comment (kingcre) [reply x/exchange/keeper/proposal_handler.go]: Thanks, I'll add those.

File: x/exchange/keeper/proposal_handler.go (hunk 7 above)
Comment (kingcre) [reply x/exchange/keeper/proposal_handler.go]: Nice catch, thanks.

File: x/exchange/simulation/proposals.go (hunk 8 above)
Comment (kingcre) [reply x/exchange/simulation/proposals.go]: OK!

File: x/exchange/types/proposal.go (hunk 9 above)
Comment (kingcre) [reply x/exchange/types/proposal.go]: Refactored code and added those checks

File: x/exchange/types/market.go (hunk 10 above)
Comment (kingcre) [reply x/exchange/types/market.go]: Added this check

File: x/exchange/types/keys.go (hunk 11 above)
Comment (kingcre) [reply x/exchange/types/keys.go:173]: As I explained earlier in morning, the role of `signByte` is to prioritize negative ticks over positive ticks in KVStore.

File: x/exchange/keeper/order.go (hunk 14 above)
Comment (kingcre) [reply x/exchange/keeper/order.go:130]: Added the check

File: x/exchange/keeper/order.go (hunk 15 above)
Comment (kingcre) [reply x/exchange/keeper/order.go]: Applied this

File: x/exchange/keeper/swap.go (hunk 18 above)
Comment (kingcre) [reply x/exchange/keeper/swap.go:89]: Markets with no limit orders will naturally be passed in the actual swap algorithm.

File: x/exchange/types/escrow.go (hunk 19 above)
Comment (kingcre) [reply x/exchange/types/escrow.go]: Negative coin amount and positive coin amount are handled separately in `Pays` and `Receives`.

File: x/exchange/types/escrow.go (hunk 20 above)
Comment (kingcre) [reply x/exchange/types/escrow.go]: Ditto

File: x/exchange/types/market.go (hunk 21 above)
Comment (kingcre) [reply x/exchange/types/market.go:64]: Added these checks

File: x/exchange/types/matching_test.go (hunk 22 above)
Comment (kingcre) [reply x/exchange/types/matching_test.go]: Nope, updated the code.

File: x/exchange/types/matching_test.go (hunk 23 above)
Comment (kingcre) [reply x/exchange/types/matching_test.go:58]: You're right. Fixed the code.

File: x/exchange/types/matching.go (hunk 24 above)
Comment (kingcre) [reply x/exchange/types/matching.go:145]: It's a good idea. I think the overhead will be small enough for those sanity checks.

File: x/exchange/types/mem_order_test.go (hunk 26 above)
Comment (kingcre) [reply x/exchange/types/mem_order_test.go:94]: Added these checks

File: x/exchange/types/mem_order.go (hunk 27 above)
Comment (kingcre) [reply x/exchange/types/mem_order.go]: That's right. It was my mistake

File: x/exchange/types/mem_order.go (hunk 28 above)
Comment (kingcre) [reply x/exchange/types/mem_order.go:105]: Good point.

File: x/exchange/keeper/genesis_test.go (hunk 12 above)
Comment (kingcre) [reply x/exchange/keeper/genesis_test.go:18]: Updated.

File: x/exchange/keeper/proposal_handler_test.go (hunk 16 above)
Comment (kingcre) [reply x/exchange/keeper/proposal_handler_test.go:54]: Wrong fees will be rejected before reaching the proposal handler, by `ValidateBasic`.

**Comment by richard-bachman** [comment]
@kingcre In L47 of `exchange/keeper/batch.go`, we have the following now.
```
if !matched {
	return
}
```
But, due to the code in L20 as follow,
```
if bestBuyPrice.LT(bestSellPrice) {
	return nil
}
```
we might not have the case of `!matched`. Is this for sanity check? Or, is there other case of not matching?

**Comment by kingcre** [comment]
> @kingcre In L47 of `exchange/keeper/batch.go`, we have the following now.
> 
> ```
> if !matched {
> 	return
> }
> ```
> 
> But, due to the code in L20 as follow,
> 
> ```
> if bestBuyPrice.LT(bestSellPrice) {
> 	return nil
> }
> ```
> 
> we might not have the case of `!matched`. Is this for sanity check? Or, is there other case of not matching?

Yes, as you said `matched` should always be `true`. It'd be better to transform it into a sanity check. Thanks.

**Comment by kingcre** [comment]
Closing. See #223
